spec.bs

```diff
@@ -223,6 +227,48 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
+  1. Let |pathname| be null.
+  1. Let |search| be null.
+  1. Let |hash| be null.
+  1. Let |processedInit| be the result of [=process a URLPatternInit=] given |init|, "`pattern`", |protocol|, |username|, |password|, |hostname|, |port|, |pathname|, |search|, and |hash|.
+  1. Set |protocol| to |processedInit|["{{URLPatternInit/protocol}}"].
+  1. Set |username| to |processedInit|["{{URLPatternInit/username}}"].
+  1. Set |password| to |processedInit|["{{URLPatternInit/password}}"].
+  1. Set |hostname| to |processedInit|["{{URLPatternInit/hostname}}"].
+  1. Set |port| to |processedInit|["{{URLPatternInit/port}}"].
+  1. Set |pathname| to |processedInit|["{{URLPatternInit/pathname}}"].
+  1. Set |search| to |processedInit|["{{URLPatternInit/search}}"].
+  1. Set |hash| to |processedInit|["{{URLPatternInit/hash}}"].
+  1. If |protocol| is a [=special scheme=] and |port| is its corresponding [=url/port=], then set |port| to the empty string.
+  1. Let |protocol component| be the result of running [=compile a component=] given |protocol|, [=canonicalize a protocol=], and [=default options=].
+  1. Let |username component| be the result of running [=compile a component=] given |username|, [=canonicalize a username=], and [=default options=].
+  1. Let |password component| be the result of running [=compile a component=] given |password|, [=canonicalize a password=], and [=default options=].
+  1. Let |hostname component| be the result of running [=compile a component=] given |hostname|, [=canonicalize a hostname=], and [=hostname options=].
+  1. Let |port component| be the result of running [=compile a component=] given |port|, [=canonicalize a port=], and [=default options=].
+  1. Let |pathname component| be null.
+  1. If the result of running [=protocol component matches a special scheme=] given |protocol component|, then set |pathname component| to the result of running [=compile a component=] given |pathname|, [=canonicalize a standard pathname=], and [=standard pathname options=].
+  1. Else set |pathname component| to the result of running [=compile a component=] given |pathname, [=canonicalize a cannot-be-a-base-URL pathname=], and [=default options=]
+  1. Let |search component| be the result of running [=compile a component=] given |search|, [=canonicalize a search=], and [=default options=].
+  1. Let |hash component| be the result of running [=compile a component=] given |hash|, [=canonicalize a hash=], and [=default options=].
+  1. Return a new {{URLPattern}} whose [=URLPattern/protocol component=] is |protocol component|, [=URLPattern/username component=] is |username component|, [=URLPattern/password component=] is |password component|, [=URLPattern/hostname component=] is |hostname component|, [=URLPattern/port component=] is |port component|, [=URLPattern/pathname component=] is |pathname component|, [=URLPattern/search component=] is |search component|, and [=URLPattern/hash component=] is |hash component|.
```

Comment:
As of about a year ago constructors no longer return new objects; instead they mutate `[=this=]`. So you want something like `Set [=this=]'s [=URLPattern/protocol component=] to ...`

You could probably remove the intermediate variables _protocol component_ etc. by just setting the appropriate field of `[=this=]` in steps 259 onward.

Reply:
I implemented your suggestion, but do I need to change the part that says things like [=URLPattern/protocol component=] "must be set upon creation"?  Its a bit weird I'm setting them in this algorithm when they have an undefined value.

Reply:
I think that is still good; the phrase isn't super well-defined but the intention is that whenever a new URLPattern is created, those things need to be set before any JS code could read them (or cause the spec to read them).